### ISSUE TITLE
Compatibility with various babel plugins

### ISSUE THREAD

**Issue by ghost** [skorlir/ava-jspm-loader#2]
​

**Comment by kdex** [comment]
Once I try removing the plugin, just to see if it the particular plugin is the problem, the error pattern continues:

```bash
╭─kdex@map-game ~/srv  ‹master*› 
╰─$ gulp cover                                                                                                                                        1 ↵
[11:06:01] Requiring external module babel-register
(node:9442) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
[11:06:02] Using gulpfile ~/srv/gulpfile.babel.js
[11:06:02] Starting 'cover'...
/home/kdex/srv/node_modules/babel-core/lib/transformation/file/index.js:591
      throw err;
      ^

TypeError: /home/kdex/srv/packages/github/jspm/nodelibs-path@0.2.0-alpha/path.js: path.resolve is not a function
    at splitPath (/home/kdex/srv/node_modules/find-up/index.js:7:14)
    at Function.module.exports.sync (/home/kdex/srv/node_modules/find-up/index.js:39:14)
    at Function.module.exports.sync (/home/kdex/srv/node_modules/read-pkg-up/index.js:21:18)
    at TestExclude.pkgConf (/home/kdex/srv/node_modules/test-exclude/index.js:52:25)
    at new TestExclude (/home/kdex/srv/node_modules/test-exclude/index.js:18:23)
    at module.exports (/home/kdex/srv/node_modules/test-exclude/index.js:88:10)
    at nycShouldInstrument (/home/kdex/srv/node_modules/babel-plugin-__coverage__/lib/index.js:38:41)
    at skip (/home/kdex/srv/node_modules/babel-plugin-__coverage__/lib/index.js:81:25)
    at PluginPass.enter (/home/kdex/srv/node_modules/babel-plugin-__coverage__/lib/index.js:467:15)
    at newFn (/home/kdex/srv/node_modules/babel-traverse/lib/visitors.js:301:19)
    at NodePath._call (/home/kdex/srv/node_modules/babel-traverse/lib/path/context.js:76:18)
    at NodePath.call (/home/kdex/srv/node_modules/babel-traverse/lib/path/context.js:48:17)
    at NodePath.visit (/home/kdex/srv/node_modules/babel-traverse/lib/path/context.js:106:12)
    at TraversalContext.visitQueue (/home/kdex/srv/node_modules/babel-traverse/lib/context.js:167:16)
    at TraversalContext.visitSingle (/home/kdex/srv/node_modules/babel-traverse/lib/context.js:118:19)
    at TraversalContext.visit (/home/kdex/srv/node_modules/babel-traverse/lib/context.js:211:19)
    at Function.traverse.node (/home/kdex/srv/node_modules/babel-traverse/lib/index.js:161:17
```

It looks like whatever it's trying to require, it's reading `undefined` fields.

**Comment by fluffywaffles** [comment]
I do think I know what is happening. Can you paste your ava config out of
package.json for me?

**Comment by kdex** [comment]
Sure. It currently looks like so:

```json
  "ava": {
    "babel": {
      "presets": [
        "es2015",
        "stage-0"
      ]
    },
    "require": [
      "babel-polyfill"
    ]
  },
  "nyc": {
    "include": [
      "src/**/*.js"
    ],
    "sourceMap": false,
    "instrument": false,
    "require": [
      "babel-register",
      "ava-jspm-loader"
    ],
    "exclude": [
      "dist",
      "components",
      "packages",
      "node_modules"
    ],
    "all": false
  }
```

**Comment by fluffywaffles** [comment]
... Huh. I notice that ava-jspm-loader is in the require array for nyc, and
not for ava.

I have only used it with ava as of now. I have never used nyc. I can look
into how it might be configurable to work with nyc. It would be helpful if
you could maybe upload a minimal test setup to a git repo - could you do
that?

**Comment by kdex** [comment]
It seems I accidentally added `ava-jspm-loader` to both `ava` and `nyc` for experimenting, but it doesn't really seem to make a difference if it's in the `require` array for `nyc` or not.

Please check out [this test case](https://github.com/kdex/ava-jspm-loader-nyc).

**Comment by fluffywaffles** [comment]
Thanks! I'll look it over.

**Comment by kdex** [comment]
Whoops, the test case was missing `ava-jspm-loader`. Make sure you pull.
Also, I came across an additional issue in the test scenario that doesn't even involve `nyc`, but only `ava` and `ava-jspm-loader`. So, maybe the issue isn't with `nyc` after all?

If you run `$(npm bin)/ava`, you will get:

```bash
╭─kdex@punchy ~/dev/ava-jspm-loader-nyc ‹master*› 
╰─$ $(npm bin)/ava
module.js:442
    throw err;
    ^

Error: Cannot find module '/home/kdex/dev/ava-jspm-loader-nyc/packages/github/jspm/nodelibs-fs@0.2.0-alpha'
    at Function.Module._resolveFilename (module.js:440:15)
    at Module._load (module.js:388:25)
    at Function.Module._load (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/ava-jspm-loader/jspm-loader.js:71:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/ava/lib/throws-helper.js:2:10)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/ava-jspm-loader/jspm-loader.js:71:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/ava/lib/test-worker.js:66:20)
    at Module._compile (module.js:541:32)

   1 exception


   ✖ test/server/util.test.js exited with a non-zero exit code: 1
```

And if you run `npm test` (which just runs `nyc --cache ava`), you will get:

```bash
╭─kdex@punchy ~/dev/ava-jspm-loader-nyc ‹master*› 
╰─$ npm test                                                                                                                                          1 ↵

> example@1.0.0 test /home/kdex/dev/ava-jspm-loader-nyc
> nyc --cache ava

/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:204
          throw new ReferenceError(messages.get("pluginUnknown", plugin, loc, i, dirname));
          ^

ReferenceError: Unknown plugin "module-alias" specified in "/home/kdex/dev/ava-jspm-loader-nyc/.babelrc" at 0, attempted to resolve relative to "/home/kdex/dev/ava-jspm-loader-nyc"
    at /home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:204:17
    at Array.map (native)
    at Function.normalisePlugins (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:180:20)
    at OptionManager.mergeOptions (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:300:36)
    at OptionManager.addConfig (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:232:10)
    at OptionManager.findConfigs (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:436:16)
    at OptionManager.init (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-core/lib/transformation/file/options/option-manager.js:484:12)
    at compile (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-register/lib/node.js:112:69)
    at loader (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-register/lib/node.js:158:14)
    at Object.require.extensions.(anonymous function) [as .js] (/home/kdex/dev/ava-jspm-loader-nyc/node_modules/babel-register/lib/node.js:168:7)
npm ERR! Test failed.  See above for more details.
```

**Comment by kdex** [comment]
Sorry, I just noticed that the test case had a few issues (I forgot to provide `ava-jspm-loader` and had an outdated jspm beta configuration).

Please pull again before looking into this further.

Note that I was able to reproduce this even without `nyc` by just running: `$(npm bin)/ava`
(So, it's likely that this is not a `nyc` issue after all.)

**Comment by fluffywaffles** [comment]
I also noticed that the error seems to be independent of `nyc`.

Here is what I found:

### The plugin

I cannot get ava-jspm-loader to play nicely with this _particular_ babel plugin. Perhaps you could use SystemJS paths as an alternative? In the example given, at least, I am able to faithfully replicate the intended behavior using `jspm.config.js`.

Otherwise, I apologize, but it appears to me that supporting this plugin is outside of the scope of the module at this time. I tried several approaches, but there seems to be some disagreement between the SystemJS loader and this babel plugin that is outside of my power to resolve.

### Without the plugin

I am able to get the loader to work with your project, sans babel-plugin-module-alias. I added the following to `jspm.config.js` in the `nodeConfig.map` section:

``` json
{
  "client/": "src/client/"
}
```

I did have to publish a patch to ava-jspm-loader to fix a bug in the way it was trying to resolve imports that did not end in a slash, but were nevertheless maintained by JSPM. The new version is v1.0.3.

Once I did that, however, everything works correctly. You can see for yourself at [skorlir/ava-jspm-loader-nyc](https://github.com/skorlir/ava-jspm-loader-nyc), my fork of your example.

**Comment by kdex** [comment]
Interesting; did the build instructions change?

When executing `npm install; jspm install; npm test` on your repository, I get:

```bash
╭─kdex@punchy ~/dev/ava-jspm-loader-nyc-2 ‹master*› 
╰─$ npm test

> example@1.0.0 test /home/kdex/dev/ava-jspm-loader-nyc-2
> nyc --cache ava

/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babel-core/lib/transformation/file/index.js:591
      throw err;
      ^

SyntaxError: /home/kdex/dev/ava-jspm-loader-nyc-2/packages/npm/buffer@4.7.0/test/node/test-buffer-alloc.js: 'return' outside of function (3:32)
  1 | 'use strict';
  2 | var Buffer = require('../../').Buffer;
> 3 | if (Buffer.TYPED_ARRAY_SUPPORT) return;
    |                                 ^
  4 | var common = {};
  5 | var assert = require('assert');
  6 | 
    at Parser.pp.raise (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/location.js:22:13)
    at Parser.pp.parseReturnStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:321:10)
    at Parser.pp.parseStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:107:19)
    at Parser.parseStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/plugins/flow.js:30:22)
    at Parser.pp.parseIfStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:314:26)
    at Parser.pp.parseStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:105:19)
    at Parser.parseStatement (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/plugins/flow.js:30:22)
    at Parser.pp.parseBlockBody (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:529:21)
    at Parser.pp.parseTopLevel (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/statement.js:36:8)
    at Parser.parse (/home/kdex/dev/ava-jspm-loader-nyc-2/node_modules/babylon/lib/parser/index.js:129:19)
npm ERR! Test failed.  See above for more details.
```

**Comment by fluffywaffles** [comment]
Ah. I was running ava directly.

Notice however that the stack trace does not refer to Module._load or to
ava-jspm-loader. I think it is unrelated. I am willing to look at it later,
however.

**Comment by fluffywaffles** [comment]
You know what it may be? Try adding "!packages/*_/_" to the "files" array
in packages.json's ava field. It looks to me like ava is trying to run
tests in the packages directory and it shouldn't.

**Comment by fluffywaffles** [comment]
I ran into something similar with node_modules. For whatever reason, ava
was not ignoring test files in node_modules either.

**Comment by kdex** [comment]
Adding `"!packages/**/*"` to `ava.files` doesn't seem to make a difference. Currently, I can only run the tasks if I enter `$(npm bin)/ava` directly.

However, issuing `$(npm bin)/nyc $(npm bin)/ava` is what results in the error above.

I just tried adding `"plugins": ["transform-minify-booleans"]` to `.babelrc`, just to see if `ava-jspm-loader` might have a problem with plugins in general, but that one seemed to work.

**Comment by fluffywaffles** [comment]
Yes, I have used ava-jspm-loader with a handful of plugins and this
babel-plugin-module-alias is the first I've ever had any trouble with.

I'm curious why nyc is erroring out. Like you, I can't help but think there
might be some interplay with ava-jspm-loader, but perhaps the problem is
simpler. Are we sure that nyc is configured correctly?

Like I said, I have never used nyc before. It would be worth it though to
double check the configuration I think.

**Comment by fluffywaffles** [comment]
@kdex check my fork for the most recent update.

You have to configure `nyc` to ignore `packages/**/*`, which otherwise (I'm assuming because they are dependencies of scripts in `src/**/*`?) it will attempt to calculate coverage for.

I am now able to run `npm test` without any trouble.

**Comment by kdex** [comment]
@skorlir Ah, this is great news!

I've been removing `babel-plugin-module-alias` from my project in the mean time and moved all the mapping to jspm (which of course can only work one-way, sine the server does not use jspm).

Thanks so much for looking into this. This solution will do.

**Comment by fluffywaffles** [comment]
Thanks @kdex for opening the issue! Glad to help.

**Comment by kdex** [comment]
@skorlir BTW: It might be a good idea to briefly mention the required `nyc` setup in the jspm recipe for `ava`. I feel like this is something that many more coverage testers might trip over. :smile:

**Comment by fluffywaffles** [comment]
I struggle a bit with that suggestion, because with JSPM you can map things
however you like, and so the exact instructions are going to differ
case-by-case... and configuration by configuration... and tool by tool.

Keeping the recipe to the point (ie, make JSPM imports work in AVA tests)
is probably better. In this case, what you're suggesting is more of a
JSPM+nyc recipe, because regardless of whether you are using AVA this issue
could arise. I don't know if the nyc repo has recipes, but if you're
interested I would open a PR over there about integrating with JSPM.

**Comment by kdex** [comment]
Well, the real issue should actually not be documenting these sorts of traps, but rather fixing all of this behavior that users might not expect. For instance, I'm still not sure why the `ava` configuration would require `"!node_modules/**/*"` when the pattern `"test/**/*"` doesn't include it, and that *might* even be unintended behavior.

Having a `jspm` setup documented over at `nyc` sounds nice. If it's not too much of a burden to you, I'd really appreciate that! :+1:

**Comment by fluffywaffles** [comment]
You bring up an interesting point, what with AVA suddenly trying to run
tests in node_modules when used with my plugin (well, it's more of a hack,
I guess, since AVA has no real plugin support of which to speak).
Strangely, I've not seen any project except yours where this happens.
Probably most npm modules put their tests in an .npmignore and so this just
doesn't happen?

Nevertheless, I agree, it seems likely that you shouldn't have to configure
that case yourself. Sounds like a bug in AVA, which somehow materializes
when using ava-jspm-loader.

**Comment by kdex** [comment]
@skorlir I've just come across yet another babel plugin that seems incompatible, showing the same `undefined` sorts of bugs: You might want to have a look at [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul).

**Comment by fluffywaffles** [comment]
@kdex how would you feel about an "exclude" feature you can set that tells ava-jspm-loader to exclude certain modules from jspm lookup?

**Comment by fluffywaffles** [comment]
It could match certain patterns by default, such as "babel-*"

**Comment by kdex** [comment]
That might indeed help, but certainly needs testing (i.e. "is the bug really here?" and "will this support babel plugins?"). Is a global `babel-*` a wise choice as a default setting?

For me, errors that start with something like

```js
/home/kdex/srv/node_modules/babel-core/lib/transformation/file/options/option-manager.js:170
        throw new TypeError(messages.get("pluginNotFunction", loc, i, typeof plugin === "undefined" ? "undefined" : (0, _typeof3.default)(plugin)));
        ^

TypeError: Plugin 0 specified in "/home/kdex/srv/.babelrc.env.test" was expected to return a function but returned "undefined"
    at Function.normalisePlugin (/home/kdex/srv/node_modules/babel-core/lib/transformation/file/options/option-manager.js:170:15)
    at /home/kdex/srv/node_modules/babel-core/lib/transformation/file/options/option-manager.js:208:30
    at Array.map (native)
    at Function.normalisePlugins (/home/kdex/srv/node_modules/babel-core/lib/transformation/file/options/option-manager.js:180:20)
…
```
could suggest that there is something wrong with how `ava-jspm-loader` handles non-jspm look-ups.

By the way, the ultimate goal is really just getting coverage testing to work in ES2015. Currently, this is still a hard problem which `nyc` tries to tackle with `babel-plugin-istanbul`.

Another plugin to test `ava-jspm-loader` against would be [`babel-plugin-__coverage__`](https://github.com/dtinth/babel-plugin-__coverage__) which, IIRC, didn't work with `ava-jspm-loader` either.

**Comment by kdex** [comment]
(I'm re-opening this issue and changing the title, as looking into this further has shown that there might be many more babel plugins that lead to this sort of crash.)

**Comment by kdex** [comment]
By the way, is `pjson.jspm.name` supposed to be defined? I noticed that for me it's not, and I can't remember if I ever had a `name` property in the `jspm` part of `package.json`.

Another thing I've noticed is that using `jspm.import` as described [here](https://github.com/avajs/ava/pull/941#issuecomment-229185405) fixes all of these issues we're having.

Is there any way that `ava-jspm-loader` could wrap this behavior with `sync`? Currently with this approach, importing is still ugly as it requires at least one serial test:

```js
const cachePromise = System.import("client/cache");
let cache;
test.serial(async t => {
	cache = (await cachePromise).default;
});
```

**Comment by fluffywaffles** [comment]
- `pjson.jspm.name`: if you ran `jspm init` and gave any name whatsoever to your app, then yes, it should be defined. If it's not defined, it's possible ava-jspm-loader will throw some kind of error, but my guess is it just doesn't add anything to the configuration. The reason for that line is to not accidentally try to require typescript files, for instance, even if jspm is configured in such a way that ordinarily it would.
- using `jspm.import`: sure, if you want to import your dependencies inside of your tests, and import them multiple times (I'm sure it would hit some kind of internal jspm cache) if multiple tests need that dependency.
- using `sync`: you _cannot_ do asynchronous work inside of Module._load, which is how ava-jspm-loader works - and you can't change Module._load to be an async function either. You're suggesting that I create some kind of library to be imported into your tests and used to decrease boilerplate. Which is fine, but that's not what this module is, and I don't prefer that solution strongly enough to convert this module to work that way. Moreover, if you already have async/await, sync does not give you anything. I'm unsure why you would try to use it with es6.

I am not interested in writing a library to be used in your tests. I am interested, however, in possibly pursuing a configuration-based approach to extending _this_ module.

For instance, the "exclude" list I mentioned before. You could also have a custom system configuration map that the loader could pass in to provide overrides to your jspm configuration. So forth. In this way, it should be possible to configure ava-jspm-loader to work regardless of the situation or plugin or module, without having to somehow incorporate all those cases into the core of ava-jspm-loader. And moreover, it won't require any additional boilerplate in your tests.

**Comment by fluffywaffles** [comment]
Just to clarify: the original issue was getting nyc to work correctly with ava and ava-jspm-loader, and we got that working, right @kdex ?

We're now just talking about babel plugins breaking.

**Comment by fluffywaffles** [comment]
The issue with babel plugins seems to stem directly from the jspm `nodelibs` shim libraries for handling the resolution of node library modules. Libraries like `fs` are not resolving correctly because jspm treats them as a special case for the browser.

Please try the fix I just pushed to the master branch.

**Comment by kdex** [comment]
@skorlir Well, not quite; the original issue was about `ava-jspm-loader` breaking for a certain babel plugin. Later on, we've discovered that it's easy to find more babel plugins for which it breaks, and that issue hasn't been resolved yet. Instead of resolving it, all we did so far was to remove one of these particular babel plugins from `.babelrc` to get it to work, and then figure out that `nyc` needs special configuration if we're dealing with `jspm` dependencies.

Now in order for `nyc` to work properly (i.e. measure accurate code coverage), other babel plugins are still needed, and unfortunately, one of those would be `babel-plugin-istanbul`. In other words: Yes, I *can* measure code coverage without it, but the numbers will be off by a lot (which really defies the whole purpose of testing code coverage).

I do understand that `Module._load` is inherently synchronous while `jspm.import` is inherently asynchronous, but that's, at least to my understanding, the whole purpose of the `sync` module, isn't it? At first glance, it seems like it provides a wrapper function, wrapping asynchronous code in a thread and then blocking it until the callback occurs.

So the intent was to wrap the asynchronous behavior of `jspm.import` into a synchronous function with the `sync` module, thus being able to rely on jspm entirely without messing with some of its internals.

> I am not interested in writing a library to be used in your tests. I am interested, however, in possibly pursuing a configuration-based approach to extending this module.

This isn't about me, really, I was just referring to the (dis)ability to use the `import` syntax with @dak's approach and how a fix for that might be crafted.

At this moment, I'm not sure to what extent a configuration will (or should) be necessary. The sole purpose of `ava-jspm-loader` is to load jspm dependencies, as configured by `jspm.config.js` and `package.json` already, so that's where the configuration already resides. Maybe I just fail to see the big picture here, but what's the purpose of, say, exclusions?

About the fix: Will test it and report my results in a few minutes. :+1: Thanks!

**Comment by kdex** [comment]
Ah, great work, @skorlir! Thanks so much!

I've briefly tested all of the three modules that we've known would break, and I see no crashes here anymore. :tada: The only issue I'm seeing now is that this might also have broken React (which depends on `fbjs`, which *is* in fact installed in my `packages`, but can't be resolved by `ava-jspm-loader`).

The file itself exists, though:

```bash
$ file packages/npm/fbjs@0.8.3/lib/invariant.js
packages/npm/fbjs@0.8.3/lib/invariant.js: ASCII text
```

Hope you don't mind a full stack trace:

```bash
Error: Cannot find module 'fbjs/lib/invariant'
    at Function.Module._resolveFilename (module.js:440:15)
    at Module._load (module.js:388:25)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/packages/npm/react@15.2.0/lib/PooledClass.js:15:17)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at loader (/home/kdex/srv/node_modules/babel-register/lib/node.js:158:5)
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:168:7)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:16:3)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/packages/npm/react@15.2.0/lib/ReactChildren.js:14:19)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at loader (/home/kdex/srv/node_modules/babel-register/lib/node.js:158:5)
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:168:7)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:16:3)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/packages/npm/react@15.2.0/lib/React.js:15:21)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at loader (/home/kdex/srv/node_modules/babel-register/lib/node.js:158:5)
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:168:7)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:16:3)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/packages/npm/react@15.2.0/react.js:3:18)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at loader (/home/kdex/srv/node_modules/babel-register/lib/node.js:158:5)
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:168:7)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:16:3)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/source/index.js:48:2)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at loader (/home/kdex/srv/node_modules/babel-register/lib/node.js:158:5)
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:168:7)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:16:3)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/test/client/index.test.js:3:1)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at extensions.(anonymous function) (/home/kdex/srv/node_modules/require-precompiled/index.js:13:11)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/ava/lib/test-worker.js:91:3)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Module._load (module.js:409:3)
    at Function.Module._load (/home/kdex/srv/node_modules/ava-jspm-loader/jspm-loader.js:80:10)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/kdex/srv/node_modules/ava/lib/test-worker.js:95:1)
    at Module._compile (module.js:541:32)
    at Module.replacementCompile (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:58:13)
    at module.exports (/home/kdex/srv/node_modules/nyc/node_modules/default-require-extensions/js.js:8:9)
    at /home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4
    at require.extensions.(anonymous function) (/home/kdex/srv/node_modules/babel-register/lib/node.js:166:7)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/node_modules/append-transform/index.js:62:4)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Function.Module.runMain (module.js:575:10)
    at Function.runMain (/home/kdex/.node-spawn-wrap-15443-144703388d95/node:40:10)
    at Object.<anonymous> (/home/kdex/srv/node_modules/nyc/bin/wrap.js:19:4)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Function.Module.runMain (module.js:575:10)
    at Object.<anonymous> (/home/kdex/.node-spawn-wrap-15443-144703388d95/node:140:8)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (node.js:348:7)
    at startup (node.js:140:9)
    at node.js:463:3
```

If this is unrelated to your fix from just now, feel free to close and track it in another issue.

**Comment by fluffywaffles** [comment]
@kdex that's an unrelated issue! Which is great news. I believe the error is coming from the fact that `ava-jspm-loader` is not checking the `jspm.config.js` "packages" config map. Rather than go down that rabbit hole, I think I have an idea for a more robust way of testing whether jspm is managing a module than the current method.

I would like to say: when I said "I am not interested in writing a library to be used in your tests," I meant "you" generically. As in, "I am not interested in writing a library to be imported and used in test files." I apologize if it sounded personal. That was not my intent.

**Comment by fluffywaffles** [comment]
I'm going to close this, but let's open a separate issue for this new bug.

**Comment by kdex** [comment]
Sure. :smile: Thanks a lot!

**Comment by kdex** [comment]
@skorlir Oh, last question: Did you intentionally not publish the fix on npm?

**Comment by fluffywaffles** [comment]
Haven't published it yet; I was waiting to hear back whether it fixed your issue. I will publish a patch now.

EDIT: published v1.0.4.

**Comment by kdex** [comment]
Thanks so much! :)
